### PR TITLE
support gamepad event  on Web && NX platform

### PR DESCRIPTION
**Pull request by PPpro**
Re: https://github.com/cocos/3d-tasks/issues/10330

## Change Log
- support gamepad input event on Web && NX platform

## Usage
```ts
// wether support gamepad event
sys.hasFeature(sys.Feature.EVENT_GAMEPAD);

// game input event
input.on(Input.EventType.GAMEPAD_INPUT, (e: EventGamepad) {
    const gamepad = e.gamepad;
    gamepad.buttonL1.getValue();  //  Button InputSource: 0-1
    gamepad.leftStick.getValue();   // Vec2 InputSource: Vec2 Object
    gamepad.leftStick.xAxis.getValue();  // Axis1D InputSource: -1-1
    gamepad.leftStick.up.getValue();  // Button InputSource: 0-1
});

// connect and disconnect gamepad
input.on(Input.EventType.GAMEPAD_CHANGE, (e) => {
    const gamepad = e.gamepad;
    console.log(gamepad.id, gamepad.connected);
});
```

## Test Case
https://github.com/cocos/cocos-test-projects/pull/613

## Tested Platform
- [x] NX platform
- [x] Web platform
- [x] WeChat platform

### CODE REVIEW

File: pal/input/minigame/gamepad-input.ts [hunk 1]
```diff
@@ -1,7 +1,124 @@
...
+import { InputEventType } from '../../../cocos/input/types/event-enum';
+import { EventTarget } from '../../../cocos/core/event/event-target';
+import { InputSourceButton, InputSourceDpad, InputSourceStick } from '../input-source';
 
-    constructor () {
-        this.support = false;
+export class GamepadInputDevice {
+    public static all: GamepadInputDevice[] = [];
+
+    public buttonNorth!: InputSourceButton;
```
Comment (minggo) [inline pal/input/minigame/gamepad-input.ts]: Public variable can be modified outside, is it any reason to use public variable?

Comment (PPpro) [reply pal/input/minigame/gamepad-input.ts]: need to change to a  public getter

File: pal/input/input-source.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
+import { Vec2, Vec3 } from '../../cocos/core/math';
+
+export abstract class InputSource<T> {
```
Comment (minggo) [inline pal/input/input-source.ts:33]: Please add docs for new classes.

File: cocos/input/types/event/event-gamepad.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+/*
+ Copyright (c) 2013-2016 Chukong Technologies Inc.
+ Copyright (c) 2017-2020 Xiamen Yaji Software Co., Ltd.
```
Comment (minggo) [inline cocos/input/types/event/event-gamepad.ts]: I think the license is out of data.

File: @types/jsb.d.ts [hunk 4]
```diff
@@ -69,6 +69,25 @@ declare namespace jsb {
...
+        value: number,
+    }
+
+    export interface ButtonInfo {
+        code: number,
+        isPressed: boolean,
+    }
+
+    export let onControllerInput: (infoList: ControllerInfo[]) => void | undefined;
+    export let onControllerChange: (controllerIds: number[]) => void | undefined;
```
Comment (minggo) [inline @types/jsb.d.ts:89]: Is it need to be exported to developers?

Comment (PPpro) [reply @types/jsb.d.ts:89]: no, it's only for pal, 
The `Input.EventType.GAMEPAD_CHANGE` is exported to developers

File: cocos/input/types/event/event-gamepad.ts (hunk 3 above)
Comment (PPpro) [reply cocos/input/types/event/event-gamepad.ts]: updated

File: pal/input/input-source.ts (hunk 2 above)
Comment (PPpro) [reply pal/input/input-source.ts:33]: added

**Review by minggo — APPROVED** [review]

**Review by SantyWang — APPROVED** [review]

**Pull request merged**